src/Emulator/Peripherals/Peripherals/Timers/STM32F4_Watchdog.cs

```diff
@@ -0,0 +1,111 @@
+//
+//  This file is licensed under the MIT License.
+//  Full license text is available in 'licenses/MIT.txt'.
+//
+
+using System;
+using Antmicro.Renode.Core;
+using Antmicro.Renode.Core.Structure.Registers;
+using Antmicro.Renode.Time;
+using Antmicro.Renode.Logging;
+
+namespace Antmicro.Renode.Peripherals.Timers
+{
+    public class STM32F4_Watchdog : BasicDoubleWordPeripheral, IKnownSize
+    {
+        //TODO: Could propably get osc_frequency from RCC.
+        //TODO: Stop timer on debug stop.
+        //TODO: Use RCC to set restart cause.
+        public STM32F4_Watchdog(Machine machine, long osc_frequency) : base(machine)
+        {
+            watchdogTimer = new LimitTimer(machine.ClockSource, osc_frequency, this, "STM32_IWDG", watchdogDefaultRLRValue, workMode: WorkMode.OneShot, enabled: false, eventEnabled: false);
+            watchdogTimer.LimitReached += TimerLimitReachedCallback;
+            DefineRegisters();
+        }
+
+        public override void Reset()
+        {
+            watchdogTimer.Reset();
+        }
+
+        public long Size => 0x400;
+        private void DefineRegisters()
+        {
+            Register.IWDG_KR.Define(this)
+            .WithValueField(0, 16, writeCallback: (_, value) =>
+            {
+                switch(value)
+                {
+                    case valueToResetWatchdog:
+                        watchdogTimer.Limit = watchdogRLRValue;
+                        registersUnlocked = false;
+                        break;
+                    case valueToStartWatchdog:
+                        watchdogTimer.EventEnabled = true;
+                        watchdogTimer.Enabled = true;
+                        registersUnlocked = false;
+                        break;
+                    case valueToUnlockRegisters:
+                        registersUnlocked = true;
+                        break;
+                    default:
+                        registersUnlocked = false;
+                        break;
+                }
+            }, name: "KEY")
+            .WithReservedBits(16, 16);
+
+            Register.IWDG_PR.Define(this)
+            .WithValueField(0, 3, writeCallback: (_, value) =>
+            {
+                watchdogTimer.Divider = (int)Math.Pow(2, (2 + value));
+            }, name: "PR")
+            .WithReservedBits(3, 29);
+
+            Register.IWDG_RLR.Define(this)
+           .WithValueField(0, 12, writeCallback: (_, value) =>
+           {
+               if(registersUnlocked)
+               {
+                   watchdogRLRValue = value;
+               }
+               else
+               {
+                   this.Log(LogLevel.Warning, "Trying to change watchdog reload value without unlocking it");
+               }
+           }, name: "RL")
+           .WithReservedBits(12, 20);
+
+            Register.IWDG_SR.Define(this)
+            .WithFlag(0, FieldMode.Read, valueProviderCallback: _ => registersUnlocked, name: "PVU")
```

Comment:
these registers should always return `false`, to be honest. The process of updating is immediate in Renode - unless you have software that actually relies on this value being non-zero?